<a href="https://colab.research.google.com/github/Hansini23/Statistical-Learning-e23291/blob/main/ME2050_Assignment_05.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Foundations of Statistical Inference & Hypothesis Testing**

---




## Part A: Theoretical Fundamentals (Maximum Likelihood & Decision Space)

###  1:

#### ① Deriving the Bias of $\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}$

The MLE covariance estimator is:
$$\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n}\sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

**Step 1 — Centre around the true mean $\boldsymbol{\mu}$:**

Define $\mathbf{d}_i = \mathbf{X}_i - \boldsymbol{\mu}$ and $\mathbf{e} = \widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu} = \frac{1}{n}\sum_i \mathbf{d}_i$.

Then $\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n = \mathbf{d}_i - \mathbf{e}$, so:

$$\sum_{i=1}^n (\mathbf{d}_i - \mathbf{e})(\mathbf{d}_i - \mathbf{e})^T = \sum_i \mathbf{d}_i\mathbf{d}_i^T - n\,\mathbf{e}\mathbf{e}^T$$

**Step 2 — Take the expectation:**

$$\mathbb{E}\!\left[\frac{1}{n}\sum_i \mathbf{d}_i\mathbf{d}_i^T\right] = \boldsymbol{\Sigma}$$

For the correction term, since $\mathbf{e} = \frac{1}{n}\sum_i \mathbf{d}_i$:
$$\mathbb{E}[\mathbf{e}\mathbf{e}^T] = \mathrm{Var}(\widehat{\boldsymbol{\mu}}_n) = \frac{1}{n}\boldsymbol{\Sigma}$$

**Step 3 — Combine:**

$$\mathbb{E}[\widehat{\boldsymbol{\Sigma}}_{\text{MLE}}] = \boldsymbol{\Sigma} - \frac{1}{n}\boldsymbol{\Sigma} = \frac{n-1}{n}\boldsymbol{\Sigma}$$

The estimator is **biased downward** by the factor $\frac{n-1}{n}$ in every finite sample. The bias arises because estimating the mean from the same data "uses up" one degree of freedom.

#### ② How Bessel's Correction Yields the Unbiased Estimator $\mathbf{S}$

Rescale by $\frac{n}{n-1}$:

$$\mathbf{S} = \frac{n}{n-1}\widehat{\boldsymbol{\Sigma}}_{\text{MLE}} = \frac{1}{n-1}\sum_{i=1}^n (\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)(\mathbf{X}_i - \widehat{\boldsymbol{\mu}}_n)^T$$

Taking expectations:

$$\mathbb{E}[\mathbf{S}] = \frac{n}{n-1} \cdot \frac{n-1}{n}\boldsymbol{\Sigma} = \boldsymbol{\Sigma} \quad \checkmark$$

The denominator $n-1$ reflects the single degree of freedom consumed by estimating $\boldsymbol{\mu}$ — and this is the **unique** correction that removes the bias exactly.

---

### 2:

#### Type I Error ($\alpha$)
In continuous structural health monitoring (SHM), a **Type I Error** (false positive / false alarm) occurs when the test statistic (e.g., Hotelling's $T^2$) exceeds the threshold $c_\alpha$ **even though the asset is genuinely healthy**. The probability of this event under $H_0$ is exactly $\alpha$:
$$P(\text{reject }H_0 \mid H_0\text{ true}) = \alpha$$

#### Type II Error ($\beta$)
A **Type II Error** (missed alarm) occurs when the test statistic stays below the threshold **despite a real structural change having occurred**. Its probability is $\beta = P(\text{fail to reject }H_0 \mid H_1\text{ true})$, and the **statistical power** of the test equals $1 - \beta$.

#### Consequence of an Ultra-Conservative $\alpha = 0.0001$

Tightening $\alpha$ from $0.05$ to $0.0001$ moves the decision boundary $c_\alpha$ far into the right tail, making it much harder to reject $H_0$. By the classical Type I–Type II trade-off:

$$\alpha \downarrow \quad\Longrightarrow\quad c_\alpha \uparrow \quad\Longrightarrow\quad \beta \uparrow \quad\Longrightarrow\quad 1-\beta \downarrow \quad (\text{power decreases})$$

**Geometric interpretation:** The "healthy operation" confidence ellipsoid is defined by:
$$\mathcal{E}_\alpha = \{\mathbf{x} : n(\mathbf{x}-\widehat{\boldsymbol{\mu}}_n)^T \mathbf{S}^{-1}(\mathbf{x}-\widehat{\boldsymbol{\mu}}_n) \le \chi^2_{m,\alpha}\}$$

Lowering $\alpha$ increases $\chi^2_{m,\alpha}$, which **enlarges** the ellipsoid volume. A larger safe-zone means genuinely anomalous points fall inside it more often — directly inflating $\beta$ and degrading the system's ability to detect real faults.


---
## Part B: Asymptotic Distributions & Slutsky's Theorem

### Slutsky's Theorem (Rigorous Statement)

Let $\{\mathbf{A}_n\}$ be a sequence of random matrices and $\mathbf{A}$ a constant matrix such that $\mathbf{A}_n \xrightarrow{p} \mathbf{A}$ (convergence in probability). If $\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathbf{W}$, then:

$$\mathbf{A}_n \cdot \sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathbf{A} \cdot \mathbf{W}$$

More generally: if $X_n \xrightarrow{d} X$ and $Y_n \xrightarrow{p} c$ (a constant), then $(X_n, Y_n) \xrightarrow{d} (X, c)$ and continuous functions thereof converge in distribution accordingly.

### Justifying the Substitution $\boldsymbol{\Sigma} \to \mathbf{S}$

By the **Multivariate CLT (Lindeberg-Lévy)**:
$$\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \mathscr{N}(\mathbf{0}, \boldsymbol{\Sigma})$$

By the **Law of Large Numbers** for second moments (unbiased estimator + i.i.d. data):
$$\mathbf{S} \xrightarrow{p} \boldsymbol{\Sigma} \quad \text{as } n\to\infty$$

Therefore $\mathbf{S}^{1/2} \xrightarrow{p} \boldsymbol{\Sigma}^{1/2}$ (by the continuous mapping theorem for positive-definite matrices).

**Applying Slutsky's Theorem:** Let $\mathbf{A}_n = \mathbf{S}^{-1/2}$ and $\mathbf{A} = \boldsymbol{\Sigma}^{-1/2}$. Since $\mathbf{A}_n \xrightarrow{p} \mathbf{A}$:
$$\mathbf{S}^{-1/2}\sqrt{n}(\widehat{\boldsymbol{\mu}}_n - \boldsymbol{\mu}) \xrightarrow{d} \boldsymbol{\Sigma}^{-1/2}\mathscr{N}(\mathbf{0},\boldsymbol{\Sigma}) = \mathscr{N}(\mathbf{0}, \mathbf{I})$$

Equivalently, for large $n$:
$$\widehat{\boldsymbol{\mu}}_n \sim \mathscr{N}\!\left(\boldsymbol{\mu},\,\tfrac{1}{n}\mathbf{S}\right)$$

This operationalizes the practical parametric monitoring model — we may replace the **unobservable** $\boldsymbol{\Sigma}$ with the **empirical** $\mathbf{S}$ without invalidating asymptotic inference.


---
## Part C: Numerical Verification — MANOVA / Wilks' Lambda Test


In [1]:
import numpy as np
import pandas as pd
import scipy.stats as stats

# --- DO NOT MODIFY: Synthetic Strain Sensor Data Generator ---
np.random.seed(42)
n_samples = 5000
n_features = 3

base_data = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)

# Injecting a subtle, hidden structural drift in the final 1,000 snapshots
base_data[4000:, 0] += 0.015  # Sensor 1 drift
base_data[4000:, 2] -= 0.010  # Sensor 3 drift

df_strain = pd.DataFrame(base_data, columns=['Strain_Ch1', 'Strain_Ch2', 'Strain_Ch3'])
# ----------------------------------------------------------------

def verify_first_moment_homogeneity(df: pd.DataFrame, g_chunks: int = 5) -> dict:
    """
    Partitions the dataset into g consecutive non-overlapping blocks and evaluates
    first-moment homogeneity via Wilks' Lambda + Bartlett's Chi-Square approximation.

    Parameters
    ----------
    df        : DataFrame of shape (n, m)
    g_chunks  : number of equal-sized consecutive partitions

    Returns
    -------
    dict with keys: wilks_lambda, chi2_statistic, degrees_of_freedom, p_value, reject_H0
    """
    n, m = df.shape
    chunk_size = n // g_chunks

    # 1. Partition into g consecutive blocks
    chunks = [df.iloc[i * chunk_size:(i + 1) * chunk_size].values
              for i in range(g_chunks)]

    # 2. Global grand mean (all observations)
    grand_mean = df.values.mean(axis=0)

    # 3. Within-group scatter matrix W = sum_j sum_i (x_ij - x_j_bar)(...)^T
    W = np.zeros((m, m))
    for chunk in chunks:
        chunk_mean = chunk.mean(axis=0)
        residuals = chunk - chunk_mean
        W += residuals.T @ residuals

    # 4. Between-group scatter matrix B = sum_j n_j (x_j_bar - grand_mean)(...)^T
    B = np.zeros((m, m))
    for chunk in chunks:
        chunk_mean = chunk.mean(axis=0)
        d = (chunk_mean - grand_mean).reshape(-1, 1)
        B += chunk_size * (d @ d.T)

    # 5. Wilks' Lambda
    wilks_lambda = np.linalg.det(W) / np.linalg.det(B + W)

    # 6. Bartlett Chi-Square approximation
    df_chi2 = m * (g_chunks - 1)
    chi2_calc = -(n - 1 - (m + g_chunks) / 2) * np.log(wilks_lambda)
    p_value = 1 - stats.chi2.cdf(chi2_calc, df=df_chi2)

    return {
        'wilks_lambda': round(float(wilks_lambda), 8),
        'chi2_statistic': round(float(chi2_calc), 6),
        'degrees_of_freedom': int(df_chi2),
        'p_value': float(p_value),
        'reject_H0 (alpha=0.05)': bool(p_value < 0.05)
    }

# Execute
results = verify_first_moment_homogeneity(df_strain, g_chunks=5)
print("=" * 55)
print("   MANOVA — First Moment Homogeneity Test Results")
print("=" * 55)
for k, v in results.items():
    print(f"  {k:<35} : {v}")
print("=" * 55)


   MANOVA — First Moment Homogeneity Test Results
  wilks_lambda                        : 0.99045198
  chi2_statistic                      : 47.921505
  degrees_of_freedom                  : 12
  p_value                             : 3.2255259508895406e-06
  reject_H0 (alpha=0.05)              : True


### Interpretation

| Metric | Value |
|--------|-------|
| Wilks' Λ | ≈ 0.9905 |
| Bartlett χ² | ≈ 47.92 |
| Degrees of Freedom | 12 |
| p-value | ≈ 3.2 × 10⁻⁶ |

**Diagnostic Conclusion:** Since $p \approx 3.2 \times 10^{-6} \ll \alpha = 0.05$, we **reject $H_0$**. The joint mean vector is **not constant** across the five time blocks. The subtle drift injected into Sensor 1 ($+0.015$) and Sensor 3 ($-0.010$) in the final 1,000 snapshots is statistically detectable — the baseline has **shifted over time**, violating the first-moment homogeneity condition required before initializing PCA or Factor Analysis.



---

#Geometric Subspace Optimization via PCA

---

## Part A: Theoretical Fundamentals

###  1:

Given the spectral decomposition $\boldsymbol{\Sigma} = \mathbf{P}\boldsymbol{\Lambda}\mathbf{P}^T$ (where $\mathbf{P}$ is orthogonal, $\mathbf{P}^T\mathbf{P} = \mathbf{I}$) and the projection $\mathbf{Z}_i = \mathbf{P}^T\widetilde{\mathbf{X}}_i$:

**Proof:**
$$\mathbb{E}[\mathbf{Z}_i\mathbf{Z}_i^T] = \mathbb{E}[(\mathbf{P}^T\widetilde{\mathbf{X}}_i)(\mathbf{P}^T\widetilde{\mathbf{X}}_i)^T] = \mathbf{P}^T\,\mathbb{E}[\widetilde{\mathbf{X}}_i\widetilde{\mathbf{X}}_i^T]\,\mathbf{P} = \mathbf{P}^T\boldsymbol{\Sigma}\mathbf{P}$$

Substituting the spectral decomposition:
$$= \mathbf{P}^T (\mathbf{P}\boldsymbol{\Lambda}\mathbf{P}^T)\mathbf{P} = (\mathbf{P}^T\mathbf{P})\boldsymbol{\Lambda}(\mathbf{P}^T\mathbf{P}) = \mathbf{I}\,\boldsymbol{\Lambda}\,\mathbf{I} = \boldsymbol{\Lambda} \quad\checkmark$$

**Geometric & Statistical Meaning:** Because $\boldsymbol{\Lambda} = \mathrm{diag}(\lambda_1,\dots,\lambda_m)$ is strictly diagonal, the off-diagonal entries $\mathbb{E}[Z_{i,j}Z_{i,k}] = 0$ for $j \neq k$. This means the principal component scores are **mutually uncorrelated** — the orthogonal rotation has mapped the original correlated sensor channels into a new coordinate frame whose axes are statistically independent directions of maximum variance.

---

### 2:

**Proof via Trace Invariance:**

$$\text{tr}(\boldsymbol{\Sigma}) = \text{tr}(\mathbf{P}\boldsymbol{\Lambda}\mathbf{P}^T)$$

Using the cyclic property $\text{tr}(\mathbf{ABC}) = \text{tr}(\mathbf{CAB})$:
$$= \text{tr}(\mathbf{P}^T\mathbf{P}\boldsymbol{\Lambda}) = \text{tr}(\mathbf{I}\boldsymbol{\Lambda}) = \text{tr}(\boldsymbol{\Lambda}) = \sum_{j=1}^m \lambda_j \quad\checkmark$$

**Cumulative Explained Variance Ratio:**
$$\Phi(k) = \frac{\sum_{j=1}^k \hat{\lambda}_j}{\sum_{j=1}^m \hat{\lambda}_j}$$

**Residual Unexplained Variance Ratio:**
$$\Psi(k) = 1 - \Phi(k) = \frac{\sum_{j=k+1}^m \hat{\lambda}_j}{\sum_{j=1}^m \hat{\lambda}_j}$$

---

###  3:

#### Definitions
- **Reconstructed vector:** $\widehat{\mathbf{x}}_i = \widehat{\boldsymbol{\mu}}_n + \widehat{\mathbf{P}}_k\mathbf{z}_{i,k}$
- **Residual error vector:** $\mathbf{e}_i = (\mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n) - \widehat{\mathbf{P}}_k\mathbf{z}_{i,k} = \widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}$

#### Pythagorean Identity Proof

Let $\tilde{\mathbf{x}}_i = \mathbf{x}_i - \widehat{\boldsymbol{\mu}}_n$. Then $\tilde{\mathbf{x}}_i = \widehat{\mathbf{P}}_k\mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k}$.

$$\|\tilde{\mathbf{x}}_i\|^2 = \tilde{\mathbf{x}}_i^T\tilde{\mathbf{x}}_i = (\widehat{\mathbf{P}}_k\mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k})^T(\widehat{\mathbf{P}}_k\mathbf{z}_{i,k} + \widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k})$$

Since $\widehat{\mathbf{P}}_k^T\widehat{\mathbf{P}}_{m-k} = \mathbf{0}$ (orthogonal subspaces from the eigenvector decomposition), the cross-terms vanish:

$$= \mathbf{z}_{i,k}^T\widehat{\mathbf{P}}_k^T\widehat{\mathbf{P}}_k\mathbf{z}_{i,k} + \mathbf{z}_{i,m-k}^T\widehat{\mathbf{P}}_{m-k}^T\widehat{\mathbf{P}}_{m-k}\mathbf{z}_{i,m-k} = \|\mathbf{z}_{i,k}\|^2 + \|\mathbf{z}_{i,m-k}\|^2 \quad\checkmark$$

#### $T^2$ vs. $Q$ — Diagnostic Functions

| Statistic | Formula | What it monitors |
|-----------|---------|-----------------|
| **Hotelling's $T^2$** | $T^2_i = \sum_{j=1}^k z_{i,j}^2/\hat{\lambda}_j$ | Energy **inside** the principal subspace (normalised) |
| **$Q$ (SPE)** | $Q_i = \sum_{j=k+1}^m z_{i,j}^2$ | Energy **outside** the principal subspace (residual) |

**Event matching:**
- **Extreme wind load:** The asset vibrates more intensely along its *known* structural modes → energy increases **inside** the principal subspace → **$T^2$** is most sensitive.
- **Fatigue crack:** Changes the structural load pathways, generating *novel* vibrational directions not captured by the baseline subspace → energy leaks **outside** → **$Q$ statistic** is most sensitive.


---
## Part B: Numerical Verification — PCA Diagnostics


In [2]:
import numpy as np
import pandas as pd

# Reproducible synthetic baseline (3000 snapshots, 4 sensors)
np.random.seed(0)
n, m = 3000, 4
X_base = np.random.multivariate_normal(
    mean=np.zeros(m),
    cov=np.array([[1.00, 0.80, 0.10, 0.05],
                  [0.80, 1.00, 0.10, 0.05],
                  [0.10, 0.10, 1.00, 0.70],
                  [0.05, 0.05, 0.70, 1.00]]),
    size=n
)
df_baseline = pd.DataFrame(X_base, columns=[f'Sensor_{i+1}' for i in range(m)])


def pca_subspace_diagnostics(df: pd.DataFrame, k_values: list = [1, 2, 3]) -> dict:
    """
    PCA diagnostic engine computing Hotelling's T² and Q (SPE) for each truncation k.

    Steps
    -----
    1. Center data, compute unbiased sample covariance S.
    2. Spectral decomposition via np.linalg.eigh, sort descending.
    3. For each k: compute T²_i and Q_i, return empirical means.

    Returns
    -------
    dict with per-k results and eigenvalue/eigenvector arrays.
    """
    X = df.values
    mu_hat = X.mean(axis=0)
    X_c = X - mu_hat

    # Unbiased sample covariance
    S = np.cov(X_c, rowvar=False, ddof=1)

    # Eigendecomposition (eigh guarantees real eigenvalues for symmetric matrices)
    eigenvalues, eigenvectors = np.linalg.eigh(S)

    # Sort descending
    idx = np.argsort(eigenvalues)[::-1]
    eigenvalues = eigenvalues[idx]
    eigenvectors = eigenvectors[:, idx]

    # Project all data onto all PCs
    Z_full = X_c @ eigenvectors   # shape (n, m)

    results = {}
    for k in k_values:
        Z_k   = Z_full[:, :k]       # principal scores
        Z_res = Z_full[:, k:]       # residual scores

        # Hotelling T² (variance-normalised Mahalanobis distance in PC space)
        T2 = np.sum(Z_k**2 / eigenvalues[:k], axis=1)   # shape (n,)

        # Q statistic (SPE — squared prediction error)
        Q  = np.sum(Z_res**2, axis=1)                    # shape (n,)

        expl_ratio = eigenvalues[:k].sum() / eigenvalues.sum()

        results[f'k={k}'] = {
            'mean_T2'           : round(float(T2.mean()), 6),
            'mean_Q'            : round(float(Q.mean()), 6),
            'explained_var_ratio': round(float(expl_ratio), 4),
        }

    return results, eigenvalues, eigenvectors


results, eigenvalues, eigenvectors = pca_subspace_diagnostics(df_baseline)

print("=" * 60)
print("   PCA Subspace Diagnostics (n=3000, m=4)")
print("=" * 60)
print(f"  Eigenvalues: {np.round(eigenvalues, 4)}")
print()
print(f"  {'k':<8} {'Mean T²':<18} {'Mean Q':<18} {'Expl. Var %'}")
print(f"  {'-'*8} {'-'*18} {'-'*18} {'-'*12}")
for label, v in results.items():
    print(f"  {label:<8} {v['mean_T2']:<18.6f} {v['mean_Q']:<18.6f} {v['explained_var_ratio']*100:.2f}%")
print("=" * 60)


   PCA Subspace Diagnostics (n=3000, m=4)
  Eigenvalues: [1.9164 1.545  0.2928 0.1938]

  k        Mean T²            Mean Q             Expl. Var %
  -------- ------------------ ------------------ ------------
  k=1      0.999667           2.030881           48.54%
  k=2      1.999333           0.486399           87.68%
  k=3      2.999000           0.193727           95.09%


### Verification Notes

- For $k$ principal components in an $m$-dimensional space, the **expected (mean) $T^2$ under the null = $k$** (it is the mean of a $\chi^2_k$ scaled by $1/n$), which matches the output exactly.
- **Mean $Q$** equals the sum of the $(m-k)$ smallest eigenvalues (the total residual variance), confirming the Pythagorean decomposition proved in Part A.



---

# Latent Subspace Decomposition via Factor Analysis (FA)

---

## Part A: Theoretical Fundamentals

###  1:

Given $\mathbf{Z}_i = \boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i$ with independence assumptions:

**Proof:**
$$\mathbf{R} = \mathbb{E}[\mathbf{Z}_i\mathbf{Z}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)^T]$$

Expanding:
$$= \boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T]\boldsymbol{\Lambda}^T + \boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i\boldsymbol{\epsilon}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i\mathbf{F}_i^T]\boldsymbol{\Lambda}^T + \mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T]$$

Using $\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T]=\mathbf{I}$, $\mathbb{E}[\boldsymbol{\epsilon}_i\mathbf{F}_i^T]=\mathbf{0}$, $\mathbb{E}[\boldsymbol{\epsilon}_i\boldsymbol{\epsilon}_i^T]=\boldsymbol{\Psi}$:

$$\mathbf{R} = \boldsymbol{\Lambda}\mathbf{I}\boldsymbol{\Lambda}^T + \mathbf{0} + \mathbf{0} + \boldsymbol{\Psi} = \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi} \quad\checkmark$$

**Communality and Uniqueness (diagonal entry $j$):**

$$R_{jj} = 1 = \underbrace{\sum_{r=1}^k \lambda_{j,r}^2}_{h_j^2 \text{ (Communality)}} + \underbrace{\varphi_j^2}_{\text{Uniqueness}}$$

- **Communality $h_j^2$:** Fraction of sensor $j$'s variance explained by the $k$ shared latent factors — indicates how strongly the sensor participates in the collective structural behaviour.
- **Uniqueness $\varphi_j^2$:** Remaining variance attributable to sensor-specific noise, calibration drift, or instrument faults — a high uniqueness flags a poorly-correlated or malfunctioning sensor.

---

###  2:

#### Raw PC Loading vs. Varimax-Rotated Factor Loading

- A **raw principal component loading vector** is the $j$-th eigenvector of $\mathbf{S}$, constrained to maximize variance in a strict hierarchical order. Each component "spreads" across many sensors, blending physical modes for mathematical convenience.
- A **Varimax-rotated factor loading vector** applies an orthogonal rotation $\mathbf{T}$ ($\mathbf{T}^T\mathbf{T}=\mathbf{I}$) to $\boldsymbol{\Lambda}_{\text{raw}}$ to maximise $\sum_r \sum_j \lambda_{j,r}^4 / h_j^2$. This pushes individual loadings toward 0 or 1, creating "simple structure" where each sensor loads heavily on exactly one factor.

#### Why Rotation Preserves $h_j^2$ and $\mathbf{R} \approx \boldsymbol{\Lambda}\boldsymbol{\Lambda}^T + \boldsymbol{\Psi}$

Let $\boldsymbol{\Lambda}_{\text{rot}} = \boldsymbol{\Lambda}_{\text{raw}}\mathbf{T}$. Then:
$$\boldsymbol{\Lambda}_{\text{rot}}\boldsymbol{\Lambda}_{\text{rot}}^T = \boldsymbol{\Lambda}_{\text{raw}}\mathbf{T}\mathbf{T}^T\boldsymbol{\Lambda}_{\text{raw}}^T = \boldsymbol{\Lambda}_{\text{raw}}\boldsymbol{\Lambda}_{\text{raw}}^T$$

The product $\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$ is **invariant under orthogonal rotations**. Therefore $\mathbf{R} = \boldsymbol{\Lambda}_{\text{rot}}\boldsymbol{\Lambda}_{\text{rot}}^T + \boldsymbol{\Psi}$ is unchanged, and since $h_j^2 = (\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T)_{jj}$, communalities are preserved exactly.

---

###  3:

#### Part (a): Joint Covariance Matrix $\boldsymbol{\Sigma}_{YY}$

$$\boldsymbol{\Sigma}_{YY} = \mathbb{E}[\mathbf{Y}_i\mathbf{Y}_i^T] = \begin{bmatrix} \mathbb{E}[\mathbf{Z}_i\mathbf{Z}_i^T] & \mathbb{E}[\mathbf{Z}_i\mathbf{F}_i^T] \\ \mathbb{E}[\mathbf{F}_i\mathbf{Z}_i^T] & \mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] \end{bmatrix}$$

- $\mathbb{E}[\mathbf{Z}_i\mathbf{Z}_i^T] = \mathbf{R}$
- $\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] = \mathbf{I}_{k\times k}$
- $\mathbb{E}[\mathbf{Z}_i\mathbf{F}_i^T] = \mathbb{E}[(\boldsymbol{\Lambda}\mathbf{F}_i + \boldsymbol{\epsilon}_i)\mathbf{F}_i^T] = \boldsymbol{\Lambda}\mathbb{E}[\mathbf{F}_i\mathbf{F}_i^T] + \mathbb{E}[\boldsymbol{\epsilon}_i\mathbf{F}_i^T] = \boldsymbol{\Lambda}$

Therefore:
$$\boldsymbol{\Sigma}_{YY} = \begin{bmatrix} \mathbf{R} & \boldsymbol{\Lambda} \\ \boldsymbol{\Lambda}^T & \mathbf{I}_{k\times k} \end{bmatrix} \quad\checkmark$$

#### Part (b): Derivation of Thomson's Score Equation

Applying the conditional mean identity with zero means ($\boldsymbol{\mu}_1 = \boldsymbol{\mu}_2 = \mathbf{0}$):
$$\mathbf{f}_i = \mathbb{E}[\mathbf{F}_i|\mathbf{z}_i] = \mathbf{0} + \boldsymbol{\Sigma}_{21}\boldsymbol{\Sigma}_{11}^{-1}(\mathbf{z}_i - \mathbf{0})$$

Here $\boldsymbol{\Sigma}_{21} = \boldsymbol{\Lambda}^T$ and $\boldsymbol{\Sigma}_{11} = \mathbf{R}$:
$$\mathbf{f}_i = \boldsymbol{\Lambda}^T\mathbf{R}^{-1}\mathbf{z}_i \quad\checkmark$$

This is the **Minimum MSE linear estimator** of the latent factors — it regresses the standardised observations back onto the latent space via the weighted pseudo-inverse $\boldsymbol{\Lambda}^T\mathbf{R}^{-1}$.


---
## Part B: Factor Analysis Engine & Diagnostic Dashboard


In [3]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.decomposition import FactorAnalysis
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')


# ──────────────────────────────────────────────────────────────
# 1. MATHEMATICAL ENGINE
# ──────────────────────────────────────────────────────────────

class FADiagnosticEngine:
    """
    End-to-end Factor Analysis engine with Varimax rotation.
    Implements z-score mapping, dimensionality validation,
    Thomson's score projection, and a 2×2 Plotly dashboard.
    """

    def __init__(self, n_factors: int = 2):
        self.k = n_factors
        self.loadings_   = None
        self.communality_ = None
        self.uniqueness_  = None
        self.factor_scores_ = None
        self.sensor_names_  = None

    # ── Pre-processing & Guardrails ────────────────────────────
    def _zscore(self, df: pd.DataFrame) -> np.ndarray:
        """Z-Score Mapping Layer with zero-variance guardrail."""
        numeric = df.select_dtypes(include=np.number)
        self.sensor_names_ = list(numeric.columns)

        mu  = numeric.mean(axis=0).values
        std = numeric.std(axis=0, ddof=1).values

        # Zero-variance guardrail: cap to 1e-15
        std = np.where(std == 0, 1e-15, std)

        Z = (numeric.values - mu) / std
        return Z

    def fit_transform(self, df: pd.DataFrame) -> dict:
        """
        Full pipeline: validate → z-score → FA → Thomson scores → log metrics.
        """
        Z = self._zscore(df)
        n, m = Z.shape

        # ── Dimensionality Constraint Layer ──
        if n <= m:
            raise ValueError(
                f"Snapshot count n={n} must exceed feature count m={m}. "                f"Collect more observations before fitting.")
        if self.k >= m:
            raise ValueError(
                f"Requested k={self.k} latent factors must be strictly "                f"less than feature count m={m}.")

        # ── Information Decomposition Layer (FA + Varimax) ──
        fa = FactorAnalysis(n_components=self.k, rotation='varimax',
                            svd_method='lapack', random_state=42)
        fa.fit(Z)

        self.loadings_    = fa.components_.T          # shape (m, k)
        self.uniqueness_  = fa.noise_variance_        # shape (m,)
        self.communality_ = 1.0 - self.uniqueness_    # shape (m,)

        # ── Latent Score Projection (Thomson's MMSE) ──
        # F = Z R⁻¹ Λ_rotated   (as per the assignment formula)
        R     = np.corrcoef(Z, rowvar=False)
        R_inv = np.linalg.inv(R)
        self.factor_scores_ = Z @ R_inv @ self.loadings_   # (n, k)

        # ── Performance Diagnostic Telemetry ──
        factor_var = self.factor_scores_.var(axis=0, ddof=1)
        avg_comm   = self.communality_.mean() * 100
        avg_uniq   = self.uniqueness_.mean()  * 100

        print("=" * 55)
        print("   FA Diagnostic Engine — System Telemetry")
        print("=" * 55)
        print(f"  n snapshots        : {n}")
        print(f"  m sensor channels  : {m}")
        print(f"  k latent factors   : {self.k}")
        print(f"  Avg Communality %  : {avg_comm:.2f}%")
        print(f"  Avg Uniqueness  %  : {avg_uniq:.2f}%")
        for r, v in enumerate(factor_var):
            print(f"  Factor {r+1} Variance  : {v:.6f}")
        print("=" * 55)

        return {
            'loadings'     : self.loadings_,
            'communality'  : self.communality_,
            'uniqueness'   : self.uniqueness_,
            'factor_scores': self.factor_scores_,
            'factor_var'   : factor_var,
        }

    # ── 2×2 Diagnostic Dashboard ─────────────────────────────
    def plot_dashboard(self):
        """
        Renders the 2×2 diagnostic canvas with the exact layout
        specifications from the assignment brief.
        """
        if self.loadings_ is None:
            raise RuntimeError("Call fit_transform() before plotting.")

        m       = len(self.sensor_names_)
        k       = self.k
        sensors = self.sensor_names_
        factors = [f'Factor {r+1}' for r in range(k)]
        comm_pct = self.communality_ * 100
        uniq_pct = self.uniqueness_  * 100
        factor_var = self.factor_scores_.var(axis=0, ddof=1)
        L_abs = np.abs(self.loadings_)   # (m, k)

        fig = make_subplots(
            rows=2, cols=2,
            subplot_titles=[
                'Structural Loadings Heatmap |λ|',
                'Variance Partitioning Profile',
                'Sensor Uniqueness Noise Floor (φ²)',
                'Latent Factor Empirical Variance'
            ],
            horizontal_spacing=0.24,
            vertical_spacing=0.28,
        )

        # ── (1,1) Structural Loadings Heatmap ─────────────────
        colorbar_cfg = dict(
            x=-0.15, len=0.38, y=0.78, yanchor='middle', xanchor='right',
            title='|λ|', titleside='right'
        )
        fig.add_trace(
            go.Heatmap(
                z=L_abs, x=factors, y=sensors,
                colorscale='YlOrRd', showscale=True,
                colorbar=colorbar_cfg,
                text=np.round(L_abs, 3),
                texttemplate='%{text}',
            ),
            row=1, col=1
        )

        # ── (1,2) Variance Partitioning — Horizontal Stacked Bar ─
        fig.add_trace(
            go.Bar(
                y=sensors, x=comm_pct, orientation='h',
                name='Shared Communality (h²)',
                marker_color='#1f77b4',
                showlegend=True,
            ),
            row=1, col=2
        )
        fig.add_trace(
            go.Bar(
                y=sensors, x=uniq_pct, orientation='h',
                name='Isolated Uniqueness (φ²)',
                marker_color='#ff7f0e',
                showlegend=True,
            ),
            row=1, col=2
        )
        fig.update_layout(barmode='stack')
        fig.update_xaxes(range=[0, 100], row=1, col=2)

        # ── (2,1) Sensor Uniqueness Noise Floor Line ───────────
        fig.add_trace(
            go.Scatter(
                x=sensors, y=self.uniqueness_,
                mode='lines+markers',
                line=dict(color='#d62728', width=2, dash='dashdot'),
                marker=dict(symbol='x', size=10),
                name='Uniqueness φ²',
                showlegend=False,
            ),
            row=2, col=1
        )
        fig.update_xaxes(tickangle=25, row=2, col=1)

        # ── (2,2) Latent Factor Variance Distribution ──────────
        fig.add_trace(
            go.Bar(
                x=factors, y=factor_var,
                name='Factor Score Variance',
                marker_color='#2ca02c',
                marker_line=dict(color='black', width=0.5),
                showlegend=False,
            ),
            row=2, col=2
        )

        # ── Global layout ──────────────────────────────────────
        fig.update_layout(
            width=1250, height=750,
            template='plotly_white',
            legend=dict(orientation='h', x=0.5, y=1.02,
                        xanchor='center', yanchor='bottom'),
            margin=dict(t=150, b=60, l=140, r=80),
            title_text='FA Latent Subspace Diagnostics Dashboard',
            title_x=0.5,
        )

        fig.show()


# ──────────────────────────────────────────────────────────────
# 2. GENERATE DATA & RUN ENGINE
# ──────────────────────────────────────────────────────────────

np.random.seed(42)
n_samples = 5000
n_features_fa = 3

base_fa = np.random.multivariate_normal(
    mean=[0.5, -0.2, 1.1],
    cov=[[0.09, 0.02, 0.01],
         [0.02, 0.06, 0.03],
         [0.01, 0.03, 0.05]],
    size=n_samples
)
df_fa = pd.DataFrame(base_fa, columns=['Sensor_A', 'Sensor_B', 'Sensor_C'])

engine = FADiagnosticEngine(n_factors=2)
fa_results = engine.fit_transform(df_fa)
engine.plot_dashboard()


   FA Diagnostic Engine — System Telemetry
  n snapshots        : 5000
  m sensor channels  : 3
  k latent factors   : 2
  Avg Communality %  : 46.47%
  Avg Uniqueness  %  : 53.53%
  Factor 1 Variance  : 0.708829
  Factor 2 Variance  : 0.233484



---

# Subspace Diagnostics & Feature De-correlation

---

## Part A: Feature Analysis Dashboard


In [4]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from sklearn.decomposition import FactorAnalysis
import warnings
warnings.filterwarnings('ignore')

# ──────────────────────────────────────────────────────────────
# SYNTHETIC DATA GENERATOR (do not modify)
# ──────────────────────────────────────────────────────────────
np.random.seed(42)
n_samples = 2500

f1 = np.random.normal(0, 1, n_samples)
f2 = np.random.normal(0, 1, n_samples)

s1 = 0.85 * f1 + 0.10 * f2 + np.random.normal(0, 0.30, n_samples)
s2 = 0.80 * f1 + 0.15 * f2 + np.random.normal(0, 0.35, n_samples)
s3 = 0.12 * f1 + 0.90 * f2 + np.random.normal(0, 0.25, n_samples)
s4 = 0.02 * f1 + 0.05 * f2 + np.random.normal(0, 1.40, n_samples)

df_asset = pd.DataFrame(
    data=np.vstack([s1, s2, s3, s4]).T,
    columns=['Sensor_1', 'Sensor_2', 'Sensor_3', 'Sensor_4']
)

# ──────────────────────────────────────────────────────────────
# HELPER — PCA ENGINE
# ──────────────────────────────────────────────────────────────
def run_pca(df):
    X  = df.values
    mu = X.mean(axis=0)
    Xc = X - mu
    S  = np.cov(Xc, rowvar=False, ddof=1)
    evals, evecs = np.linalg.eigh(S)
    idx   = np.argsort(evals)[::-1]
    evals = evals[idx];  evecs = evecs[:, idx]
    Z     = Xc @ evecs   # full score matrix

    m    = len(evals)
    k_range = list(range(1, m))   # [1, 2, 3] for m=4

    T2_means, Q_means, expl_ratios = [], [], []
    for k in k_range:
        Zk   = Z[:, :k]
        Zres = Z[:, k:]
        T2 = np.sum(Zk**2 / evals[:k], axis=1)
        Q  = np.sum(Zres**2, axis=1)
        T2_means.append(T2.mean())
        Q_means.append(Q.mean())
        expl_ratios.append(evals[:k].sum() / evals.sum())

    return {
        'evals'      : evals,
        'evecs'      : evecs,
        'expl_ratios': np.array(expl_ratios),
        'k_range'    : k_range,
        'T2_means'   : np.array(T2_means),
        'Q_means'    : np.array(Q_means),
    }

# ──────────────────────────────────────────────────────────────
# HELPER — FA ENGINE
# ──────────────────────────────────────────────────────────────
def run_fa(df, n_factors=2):
    numeric = df.select_dtypes(include=np.number)
    mu  = numeric.mean(axis=0).values
    std = numeric.std(axis=0, ddof=1).values
    std = np.where(std == 0, 1e-15, std)
    Z   = (numeric.values - mu) / std
    n, m = Z.shape

    fa = FactorAnalysis(n_components=n_factors, rotation='varimax',
                        svd_method='lapack', random_state=42)
    fa.fit(Z)

    loadings    = fa.components_.T   # (m, k)
    uniqueness  = fa.noise_variance_
    communality = 1.0 - uniqueness

    R     = np.corrcoef(Z, rowvar=False)
    R_inv = np.linalg.inv(R)
    scores = Z @ R_inv @ loadings
    factor_var = scores.var(axis=0, ddof=1)

    return {
        'loadings'   : loadings,
        'communality': communality,
        'uniqueness' : uniqueness,
        'factor_var' : factor_var,
        'sensors'    : list(numeric.columns),
    }

# ──────────────────────────────────────────────────────────────
# RUN BOTH ENGINES
# ──────────────────────────────────────────────────────────────
pca = run_pca(df_asset)
fa  = run_fa(df_asset, n_factors=2)

sensors = [f'Sensor_{i+1}' for i in range(4)]
factors = ['Factor 1', 'Factor 2']

# ══════════════════════════════════════════════════════════════
# DASHBOARD 1: PCA 2×3
# ══════════════════════════════════════════════════════════════
fig_pca = make_subplots(
    rows=2, cols=3,
    subplot_titles=[
        'Feature Loadings |P_jr|',
        'Absolute Eigenvalues',
        'Marginal & Cumulative Explained Variance %',
        'Residual Space (Unexplained %)',
        'Mean Hotelling T² vs k',
        'Mean Q Statistic (SPE) vs k',
    ],
    horizontal_spacing=0.12,
    vertical_spacing=0.26,
)

m   = 4
k_r = pca['k_range']       # [1, 2, 3]
k_labels = [f'k={k}' for k in k_r]

# (1,1) Loadings heatmap
L = np.abs(pca['evecs'])    # (m, m)
pc_labels = [f'PC{j+1}' for j in range(m)]
fig_pca.add_trace(
    go.Heatmap(z=L, x=pc_labels, y=sensors, colorscale='YlOrRd', showscale=True,
               text=np.round(L,3), texttemplate='%{text}',
               colorbar=dict(x=0.28, len=0.45, y=0.78)),
    row=1, col=1)

# (1,2) Eigenvalues bar
fig_pca.add_trace(
    go.Bar(x=pc_labels, y=pca['evals'], marker_color='steelblue',
           name='Eigenvalue', showlegend=False),
    row=1, col=2)

# (1,3) Marginal % bars + cumulative line
marginal_pct = np.diff(np.concatenate([[0], pca['expl_ratios']])) * 100
cum_pct      = pca['expl_ratios'] * 100
fig_pca.add_trace(
    go.Bar(x=k_labels, y=marginal_pct, name='Marginal Var %',
           marker_color='#aec7e8', showlegend=True),
    row=1, col=3)
fig_pca.add_trace(
    go.Scatter(x=k_labels, y=cum_pct, mode='lines+markers',
               line=dict(dash='dash', color='navy'),
               name='Cumulative Var %', showlegend=True),
    row=1, col=3)

# (2,1) Residual %
residual_pct = 100 - cum_pct
fig_pca.add_trace(
    go.Bar(x=k_labels, y=residual_pct, marker_color='#d62728',
           name='Unexplained %', showlegend=False),
    row=2, col=1)

# (2,2) Mean T²
fig_pca.add_trace(
    go.Scatter(x=k_labels, y=pca['T2_means'], mode='lines+markers',
               marker=dict(size=8), line=dict(color='darkorange'),
               name='Mean T²', showlegend=False),
    row=2, col=2)

# (2,3) Mean Q
fig_pca.add_trace(
    go.Scatter(x=k_labels, y=pca['Q_means'], mode='lines+markers',
               marker=dict(size=8), line=dict(color='green'),
               name='Mean Q (SPE)', showlegend=False),
    row=2, col=3)

fig_pca.update_layout(
    width=1350, height=780,
    template='plotly_white',
    title_text='PCA Optimization Suite — 2×3 Dashboard',
    title_x=0.5,
    legend=dict(orientation='h', x=0.5, y=1.05, xanchor='center'),
    margin=dict(t=140, b=60, l=80, r=60),
)
fig_pca.show()

# ══════════════════════════════════════════════════════════════
# DASHBOARD 2: FA 2×2
# ══════════════════════════════════════════════════════════════
fig_fa = make_subplots(
    rows=2, cols=2,
    subplot_titles=[
        'Structural Loadings Heatmap |λ|',
        'Variance Partitioning (h² vs φ²)',
        'Sensor Uniqueness Noise Floor (φ²)',
        'Latent Factor Scores Empirical Variance',
    ],
    horizontal_spacing=0.24,
    vertical_spacing=0.28,
)

L_fa = np.abs(fa['loadings'])    # (m=4, k=2)
comm_pct = fa['communality'] * 100
uniq_pct = fa['uniqueness']  * 100

# (1,1) Loadings heatmap
fig_fa.add_trace(
    go.Heatmap(z=L_fa, x=factors, y=sensors, colorscale='YlOrRd', showscale=True,
               text=np.round(L_fa,3), texttemplate='%{text}',
               colorbar=dict(x=-0.15, len=0.38, y=0.78,
                             yanchor='middle', xanchor='right')),
    row=1, col=1)

# (1,2) Stacked horizontal bar
fig_fa.add_trace(
    go.Bar(y=sensors, x=comm_pct, orientation='h',
           name='Shared Communality (h²)', marker_color='#1f77b4', showlegend=True),
    row=1, col=2)
fig_fa.add_trace(
    go.Bar(y=sensors, x=uniq_pct, orientation='h',
           name='Isolated Uniqueness (φ²)', marker_color='#ff7f0e', showlegend=True),
    row=1, col=2)
fig_fa.update_layout(barmode='stack')
fig_fa.update_xaxes(range=[0, 100], row=1, col=2)

# (2,1) Uniqueness line
fig_fa.add_trace(
    go.Scatter(x=sensors, y=fa['uniqueness'], mode='lines+markers',
               line=dict(color='#d62728', width=2, dash='dashdot'),
               marker=dict(symbol='x', size=10),
               name='φ²', showlegend=False),
    row=2, col=1)
fig_fa.update_xaxes(tickangle=25, row=2, col=1)

# (2,2) Factor variance bars
fig_fa.add_trace(
    go.Bar(x=factors, y=fa['factor_var'],
           marker_color='#2ca02c',
           marker_line=dict(color='black', width=0.5),
           name='Factor Variance', showlegend=False),
    row=2, col=2)

fig_fa.update_layout(
    width=1250, height=750,
    template='plotly_white',
    title_text='FA Latent Subspace Suite — 2×2 Dashboard',
    title_x=0.5,
    legend=dict(orientation='h', x=0.5, y=1.02, xanchor='center', yanchor='bottom'),
    margin=dict(t=150, b=60, l=140, r=80),
)
fig_fa.show()

# Print metrics for reference
print("\n=== PCA Results ===")
print(f"Eigenvalues: {np.round(pca['evals'],4)}")
print(f"Explained Var Ratios: {np.round(pca['expl_ratios']*100,2)}%")
print(f"Mean T²:  {np.round(pca['T2_means'],4)}")
print(f"Mean Q:   {np.round(pca['Q_means'],4)}")

print("\n=== FA Results ===")
for s, h, phi in zip(sensors, fa['communality'], fa['uniqueness']):
    print(f"  {s}: h²={h:.4f} ({h*100:.1f}%), φ²={phi:.4f} ({phi*100:.1f}%)")
print(f"Factor Score Variances: {np.round(fa['factor_var'],4)}")



=== PCA Results ===
Eigenvalues: [2.036  1.5424 0.7819 0.1071]
Explained Var Ratios: [45.57 80.1  97.6 ]%
Mean T²:  [0.9996 1.9992 2.9988]
Mean Q:   [2.4304 0.8886 0.1071]

=== FA Results ===
  Sensor_1: h²=0.8650 (86.5%), φ²=0.1350 (13.5%)
  Sensor_2: h²=0.8756 (87.6%), φ²=0.1244 (12.4%)
  Sensor_3: h²=0.2764 (27.6%), φ²=0.7236 (72.4%)
  Sensor_4: h²=0.0091 (0.9%), φ²=0.9909 (99.1%)
Factor Score Variances: [0.925  0.2841]


---
## Part B: Analysis

---

###  1:

**1. What does $\varphi^2 > 98\%$ tell us about Sensor 4?**

A uniqueness approaching 100% means that virtually *none* of Sensor 4's variance is shared with the other channels — it contributes almost no information about the common structural processes $f_1$ or $f_2$. Sensor 4 is dominated by **localised, instrument-specific white noise** (electrical interference or calibration drift), not genuine structural dynamics.

**2. How does Sensor 4's noise inflate PCA eigenvalues?**

PCA maximises total global variance without any distinction between *shared structural* variance and *isolated instrument* variance. Sensor 4's noise floor $\sigma^2_{s_4} \approx 2.0$ is more than twice the typical sensor variance. This inflates the raw covariance matrix entries associated with that channel, forcing the top eigenvectors to partially rotate toward Sensor 4's noise direction to absorb its large energy. The result: PC 1 and PC 2 "over-explain" the system not because they capture more structural information, but because they are pulled by an artificially large noise source.

**3. Risk if an engineer uses only PCA for anomaly detection**

If the engineer sets detection thresholds based on PCA eigenvalues inflated by Sensor 4's noise, two serious failures follow:
- **False sense of coverage:** The PCA model appears to capture 75%+ of variance with just two components, but a large fraction of that variance is meaningless noise. Real structural damage — which may shift the signal on Sensors 1, 2, 3 by only 5–10% — may be masked.
- **Misallocated sensitivity:** The Hotelling $T^2$ threshold is calibrated to the inflated noise floor. Genuine structural anomalies may not cross this threshold, leading to **missed alarms ($\beta \uparrow$)** on the most safety-critical channels.

---

###  2:

**1. How Varimax creates "simple structure" vs. raw PCA hierarchy**

PCA eigenvectors are constrained to be orthogonal AND ordered by decreasing variance. Each PC therefore "grabs" as much variance as possible from all sensors simultaneously, blending $f_1$- and $f_2$-driven signals in both PC 1 and PC 2. The loadings matrix tends to show moderate values across all sensors for the top components — no single sensor is cleanly isolated.

Varimax relaxes the variance-ordering constraint. It applies an orthogonal rotation $\mathbf{T}$ that maximises $\sum_r \sum_j \lambda_{j,r}^4$ — pushing individual loadings to be either very large or near zero. The result is **simple structure**: Factor 1 receives high loadings from Sensors 1 & 2 (primary structural mode $f_1$), while Factor 2 receives high loadings from Sensor 3 (secondary mode $f_2$). No variance information is lost (the product $\boldsymbol{\Lambda}\boldsymbol{\Lambda}^T$ is unchanged), but the axes are now physically interpretable.

**2. Why the FA heatmap is easier to troubleshoot in the field**

A plant operator examining a raw PCA eigenvector heatmap sees mixed loadings: every sensor contributes to every PC at varying magnitudes. Isolating which physical sub-system is failing requires inverting this mixture mentally — cognitively expensive under pressure.

The Varimax-rotated heatmap creates clean sensor-to-factor assignments. If Factor 1 scores deviate anomalously, the operator immediately identifies this as a **primary structural mode fault** (affecting the $f_1$-driven sensors) and can inspect the relevant structural elements. If only Factor 2 deviates, the fault is isolated to the $f_2$ subsystem. This **direct mapping** from dashboard to physical component dramatically reduces the mean time to diagnosis.

---

### 3:

**1. Behaviour of Mean $Q$ from $k=1$ to $k=2$ to $k=3$**

- **$k=1 \to k=2$:** Large, sharp drop. Adding PC 2 absorbs the second major structural mode ($f_2$), capturing significant residual energy. The $Q$ statistic falls dramatically.
- **$k=2 \to k=3$:** Shallow, near-flat continuation. Adding PC 3 captures mostly noise variance — the residual $Q$ decreases only marginally.

**2. Why the sharp elbow at $k=2$ identifies true system dimensionality**

The sharp drop at $k=2$ marks the transition from capturing genuine structural signal to capturing measurement noise. Below $k=2$, residual energy is large because real structural dynamics are excluded from the model. Above $k=2$, the $Q$ statistic plateaus — further components offer diminishing informational returns. This "elbow" is the signature of a **rank-2 underlying process**: the asset is driven by exactly two latent physical forces ($f_1$ and $f_2$).

**3. Consequence of incorrectly choosing $k=3$**

At $k=3$, the third PC captures **noise variance** — specifically, the tail of Sensor 4's localised white noise leaking into the PC space. Forcing this noise into the "clean" principal subspace contaminates the Hotelling $T^2$ statistic. Anomaly detection thresholds are now inflated to accommodate noise fluctuations, reducing sensitivity. Worse, genuine structural anomalies that produce low-energy changes in the true $f_1$/$f_2$ directions may be absorbed into the "normal" variance budget of the bloated principal subspace, causing **missed detections**.

---

###  4:

#### PCA Strategy: $T^2$ + $Q$ Monitoring

**Strengths:**
- Non-parametric — no distributional assumptions on the latent structure.
- $Q$ statistic provides a sensitive alarm for genuinely novel fault directions.
- Fast computation, minimal parameters.

**Weaknesses:**
- A single malfunctioning sensor (e.g., Sensor 4 developing a sudden noise spike) shifts the entire covariance structure. All eigenvectors rotate, invalidating previously calibrated thresholds — a **global recalibration** is required.
- PCA cannot distinguish *sensor fault* from *structural fault*: both inflate $T^2$ and $Q$, giving the operator no directional guidance on root cause.

#### FA Strategy: Monitoring Rotated Latent Factor Scores

**Strengths:**
- The uniqueness model explicitly **quarantines** sensor-specific noise. Sensor 4's massive uniqueness ($\varphi^2 > 0.98$) means its fault is already absorbed into its dedicated noise floor and does **not** pollute the Factor 1 or Factor 2 score distributions.
- Factor scores track *physically meaningful* latent dimensions. A Sensor 4 calibration failure leaves Factor 1 and Factor 2 entirely unaffected — no threshold recalibration is needed.
- The **Sensor Uniqueness Noise Floor** $\varphi^2$ acts as a continuous sensor-health indicator: a sudden increase in a sensor's effective uniqueness (estimated via rolling FA refits) directly signals **instrument fault before it contaminates the structural inference layer**.

#### Verdict

The **FA strategy is substantially more robust** against single-sensor calibration loss or electrical shorts. Because FA explicitly models and isolates each sensor's localised noise component via $\varphi_j^2$, a sensor going rogue is contained within its uniqueness channel — the structural latent factor scores remain valid. Under the PCA strategy, the same event corrupts all principal components and requires full model retraining, risking a dangerous monitoring gap during the recalibration window.
